server/cmd/api/api/chromium.go

```diff
@@ -0,0 +1,210 @@
+			}
+			current.name = name
+		default:
+			return oapi.UploadExtensionsAndRestart400JSONResponse{BadRequestErrorJSONResponse: oapi.BadRequestErrorJSONResponse{Message: "invalid field"}}, nil
+		}
+		// If we have both fields, finalize this item
+		if current != nil && current.zipReceived && current.name != "" {
+			items = append(items, *current)
+			current = nil
+		}
+	}
+
+	// If the last pair is incomplete, reject the request
+	if current != nil && (!current.zipReceived || current.name == "") {
+		return oapi.UploadExtensionsAndRestart400JSONResponse{BadRequestErrorJSONResponse: oapi.BadRequestErrorJSONResponse{Message: "each extension must include consecutive name and zip_file"}}, nil
+	}
+
+	if len(items) == 0 {
+		return oapi.UploadExtensionsAndRestart400JSONResponse{BadRequestErrorJSONResponse: oapi.BadRequestErrorJSONResponse{Message: "no extensions provided"}}, nil
+	}
+
+	// Materialize uploads
+	extBase := "/home/kernel/extensions"
+
+	for _, p := range items {
+		if !p.zipReceived || p.name == "" {
+			return oapi.UploadExtensionsAndRestart400JSONResponse{BadRequestErrorJSONResponse: oapi.BadRequestErrorJSONResponse{Message: "each item must include zip_file and name"}}, nil
+		}
+		dest := filepath.Join(extBase, p.name)
+		if err := os.MkdirAll(dest, 0o755); err != nil {
+			return oapi.UploadExtensionsAndRestart500JSONResponse{InternalErrorJSONResponse: oapi.InternalErrorJSONResponse{Message: "failed to create extension dir"}}, nil
+		}
+		if err := ziputil.Unzip(p.zipTemp, dest); err != nil {
+			return oapi.UploadExtensionsAndRestart400JSONResponse{BadRequestErrorJSONResponse: oapi.BadRequestErrorJSONResponse{Message: "invalid zip file"}}, nil
+		}
+		if err := exec.Command("chown", "-R", "kernel:kernel", dest).Run(); err != nil {
+			log.Error("failed to chown extension dir", "err", err)
+			return oapi.UploadExtensionsAndRestart500JSONResponse{InternalErrorJSONResponse: oapi.InternalErrorJSONResponse{Message: "failed to chown extension dir"}}, nil
+		}
+		log.Info("installed extension", "name", p.name)
+	}
+
+	// Build flags overlay file in /chromium/flags
+	var paths []string
+	for _, p := range items {
+		paths = append(paths, filepath.Join(extBase, p.name))
+	}
+	overlay := fmt.Sprintf("--disable-extensions-except=%s --load-extension=%s\n", strings.Join(paths, ","), strings.Join(paths, ","))
+	if err := exec.Command("mkdir", "-p", "/chromium").Run(); err != nil {
+		return oapi.UploadExtensionsAndRestart500JSONResponse{InternalErrorJSONResponse: oapi.InternalErrorJSONResponse{Message: "failed to create chromium dir"}}, nil
+	}
+	if err := os.WriteFile("/chromium/flags", []byte(overlay), 0o644); err != nil {
+		return oapi.UploadExtensionsAndRestart500JSONResponse{InternalErrorJSONResponse: oapi.InternalErrorJSONResponse{Message: "failed to write overlay flags"}}, nil
+	}
```

Comment:
should this be read -> write? If I'm understanding this we won't preserve any previously loaded extensions. is that desired?

Reply:
good catch... will do some flag merging now that we have code for that